src/outline.rs

```diff
@@ -34,65 +44,334 @@ impl OutlineNode {
-                    Arc::make_mut(last_child).add_child(child);
+    /// Add this node to the given tree, with the given children, and set their parent id
+    fn add_to_tree(self, tree: &mut Tree<OutlineNode>, children: &[NodeId]) -> NodeId {
+        let node_id = tree.add_node(self, children);
+        for child_node_id in children {
+            tree[*child_node_id].parent = node_id;
+        }
+        node_id
+    }
+
+    /// Returns the name of that node, i.e. the heading title or statement label
+    fn get_name<'a>(&'a self, database: &'a Database) -> &'a str {
+        let sref = database.parse_result().statement(self.stmt_address);
+        match self.level {
+            HeadingLevel::Statement => as_str(sref.label()),
+            HeadingLevel::Database => "Database",
+            _ => as_str(&self.name),
+        }
+    }
+}
+
+/// A reference to an outline node
+#[derive(Debug, Clone, Copy)]
+#[allow(missing_docs)]
+pub enum OutlineNodeRef<'a> {
+    Chapter {
+        database: &'a Database,
+        node_id: NodeId,
+    },
+    Statement {
+        database: &'a Database,
+        sref: StatementRef<'a>,
+    },
+}
+
+impl<'a> OutlineNodeRef<'a> {
+    /// Creates an `OutlineNodeRef` for the root of this database
+    #[must_use]
+    pub fn root_node(database: &'a Database) -> Self {
+        OutlineNodeRef::Chapter {
+            database,
+            node_id: database.outline_result().root,
+        }
+    }
+
+    /// Creates an `OutlineNodeRef` for the given statement
+    #[must_use]
+    pub const fn statement_node(database: &'a Database, sref: StatementRef<'a>) -> Self {
+        OutlineNodeRef::Statement { database, sref }
+    }
+
+    /// Returns this node's parent, or `None` if this is the root node
+    #[inline]
+    #[must_use]
+    pub fn parent(self) -> Option<OutlineNodeRef<'a>> {
+        match self {
+            OutlineNodeRef::Chapter { database, node_id } => {
+                let node = &database.outline_result().tree[node_id];
+                if node.level == HeadingLevel::Database {
+                    None
                 } else {
-                    self.children.push(Arc::new(child));
+                    Some(OutlineNodeRef::Chapter {
+                        database,
+                        node_id: node.parent,
+                    })
                 }
             }
+            OutlineNodeRef::Statement { database, sref } => {
+                let outline = database.outline_result();
+                Some(OutlineNodeRef::Chapter {
+                    database,
+                    node_id: outline.statement_node_inside(sref.address(), outline.root, database),
+                })
+            }
         }
     }
+}
 
-    /// Returns the name of that node, i.e. the heading title
+impl OutlineNodeRef<'_> {
+    /// Iterator through the children outline nodes
+    #[must_use]
+    pub fn children_iter(&self) -> Box<dyn Iterator<Item = OutlineNodeRef<'_>> + '_> {
```

Comment:
One way to make this a single iterator type with combinators is to use `Some(iter)` in the first case and `None` in the second, and after the match apply `.into_iter().flatten()`. But for something like this I would just write out the explicit state machine and implement a manual iterator, which has the advantage that the result type can be named, which is useful for consumers that want to put it in a struct or other iterator.

Reply:
Actually, looking at this more carefully I think there is an issue here with the recursive call. It's calling the recursion in the second argument of the `chain`, but both iterators are evaluated eagerly so you end up with a big linked list all at once. The manual iterator approach does not have this issue: the recursive call does not happen until one element has already been yielded.

Reply:
Actually this is not recursive, this is a single call to the iterator `children_iter` on the outline `Tree` object which happens to have the same name. Is this too confusing?

Re your first comment, Ok, I'll extract a special struct.

Reply:
I just pushed this change myself, but feel free to make any changes.

Reply:
Looks good to me. I see you did the same for the ancestor iterator, which is good too.